src/merkle/store/tests.rs

```diff
@@ -647,6 +647,149 @@ fn test_constructors() -> Result<(), MerkleError> {
+        root = store.set_node(root, index, node).unwrap().root;
+        assert_eq!(64, store.get_leaf_depth(root, k).unwrap());
+    }
+}
+
+#[test]
+fn get_leaf_depth_works_with_incremental_depth() {
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // insert some path to the left of the root and assert it
+    let key = 0b01001011_10110110_00001101_01110100_00111011_10101101_00000100_01000001_u64;
+    assert_eq!(0, store.get_leaf_depth(root, key).unwrap());
+    let depth = 64;
+    let index = NodeIndex::new(depth, key).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // flip the key to the right of the root and insert some content on depth 16
+    let key = 0b11001011_10110110_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(1, store.get_leaf_depth(root, key).unwrap());
+    let depth = 16;
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // attempt the sibling of the previous leaf
+    let key = 0b11001011_10110111_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(16, store.get_leaf_depth(root, key).unwrap());
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // move down to the next depth and assert correct behavior
+    let key = 0b11001011_10110100_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(15, store.get_leaf_depth(root, key).unwrap());
+    let depth = 17;
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+}
+
+#[test]
+fn get_leaf_depth_works_with_small_depth() {
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+    let node = [ONE; WORD_SIZE];
+
+    // insert some random, 8 depth keys. `a` diverges from the first bit
+    let a = 0b01101001_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let b = 0b10011001_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let c = 0b10010110_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let d = 0b11110110_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+
+    for k in [a, b, c, d] {
+        let index = NodeIndex::new(8, k >> 56).unwrap();
```

Comment:
if we declare keys without trailing zeros this could be just:
```Rust
let index = NodeIndex::new(8, k).unwrap();
```

Reply:
It is true, but that might be a bit confusing for people who aren't aware of the context. If we keep all bits explicit, it is unambiguous.

Let's continue the discussion here
https://github.com/0xPolygonMiden/crypto/pull/125#discussion_r1159045463